File: Python3/0901_online_stock_span.py
# https://leetcode.cn/problems/online-stock-span/

class StockSpanner:

    def __init__(self):
        self.stack = [[-1, float("inf")]]
        self.idx = -1

    def next(self, price):
        """
        Inputs:
            price: int
        Outputs:
            res: int
        """
        self.idx += 1
        while price >= self.stack[-1][1]:
            self.stack.pop()
        self.stack.append([self.idx, price])
        res = self.idx - self.stack[-2][0]
        return res

# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)

File: Cpp/0901_online_stock_span.cpp
// https://leetcode.cn/problems/online-stock-span/

class StockSpanner {
public:
    StockSpanner() {
        stk.push_back(make_pair(-1, INT_MAX));
        idx = -1;
    }
    
    int next(int price) {
        idx += 1;
        while (price >= stk.back().second) {
            stk.pop_back();
        }
        int res = idx - stk.back().first;

        stk.push_back(make_pair(idx, price));
        
        return res;
    }

private:
    vector<pair<int,int>> stk;
    int idx;
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */